File: Sources/SkillManagerCore/SkillParser.swift
import Foundation

public struct ParsedSkill: Hashable, Sendable {
    public var name: String
    public var title: String
    public var summary: String
    public var contextTokenEstimate: Int

    public init(name: String, title: String, summary: String, contextTokenEstimate: Int) {
        self.name = name
        self.title = title
        self.summary = summary
        self.contextTokenEstimate = contextTokenEstimate
    }
}

public enum SkillParser {
    public static func parse(skillMarkdownURL: URL) -> ParsedSkill? {
        guard let data = try? Data(contentsOf: skillMarkdownURL),
              let raw = String(data: data, encoding: .utf8) else {
            return nil
        }

        let folderName = skillMarkdownURL.deletingLastPathComponent().lastPathComponent
        let frontmatter = extractFrontmatter(from: raw)
        let name = frontmatter["name"]?.trimmedNonEmpty ?? folderName
        let description = frontmatter["description"]?.trimmedNonEmpty
        let heading = raw
            .components(separatedBy: .newlines)
            .first { $0.hasPrefix("# ") }?
            .dropFirst(2)
            .description
            .trimmedNonEmpty

        let summary = description ?? firstMeaningfulParagraph(in: raw) ?? "本地 Skill"
        return ParsedSkill(
            name: name,
            title: heading ?? name,
            summary: summary,
            contextTokenEstimate: estimateTokens("\(name)\n\(summary)")
        )
    }

    public static func estimateTokens(_ text: String) -> Int {
        let scalarCount = text.unicodeScalars.count
        let wordCount = text.split { $0.isWhitespace || $0.isPunctuation }.count
        let cjkCount = text.unicodeScalars.filter { scalar in
            (0x4E00...0x9FFF).contains(Int(scalar.value))
        }.count

        let latinEstimate = max(wordCount, scalarCount / 5)
        return max(1, latinEstimate + cjkCount / 2)
    }

    private static func extractFrontmatter(from raw: String) -> [String: String] {
        let lines = raw.components(separatedBy: .newlines)
        guard lines.first?.trimmingCharacters(in: .whitespacesAndNewlines) == "---" else {
            return [:]
        }

        var result: [String: String] = [:]
        for line in lines.dropFirst() {
            if line.trimmingCharacters(in: .whitespacesAndNewlines) == "---" {
                break
            }
            guard let colon = line.firstIndex(of: ":") else { continue }
            let key = String(line[..<colon]).trimmingCharacters(in: .whitespacesAndNewlines)
            var value = String(line[line.index(after: colon)...]).trimmingCharacters(in: .whitespacesAndNewlines)
            if value.hasPrefix("\""), value.hasSuffix("\""), value.count >= 2 {
                value.removeFirst()
                value.removeLast()
            }
            result[key] = value
        }
        return result
    }

    private static func firstMeaningfulParagraph(in raw: String) -> String? {
        raw.components(separatedBy: "\n\n")
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .first { paragraph in
                !paragraph.isEmpty &&
                !paragraph.hasPrefix("---") &&
                !paragraph.hasPrefix("#") &&
                paragraph.count > 16
            }
    }
}

private extension String {
    var trimmedNonEmpty: String? {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}

File: packages/core/src/skill-parser.ts
import fs from "node:fs";
import path from "node:path";

export interface ParsedSkill {
  name: string;
  title: string;
  summary: string;
  contextTokens: number;
  content: string;
}

export async function parseSkillMarkdown(skillMarkdownPath: string): Promise<ParsedSkill | null> {
  let raw: string;
  try {
    raw = await fs.promises.readFile(skillMarkdownPath, "utf8");
  } catch {
    return null;
  }

  const folderName = path.basename(path.dirname(skillMarkdownPath));
  const frontmatter = extractFrontmatter(raw);
  const name = nonEmpty(frontmatter.name) ?? folderName;
  const summary = nonEmpty(frontmatter.description) ?? firstMeaningfulParagraph(raw) ?? "Local skill";
  const heading = raw
    .split(/\r?\n/)
    .find((line) => line.startsWith("# "))
    ?.slice(2)
    .trim();

  return {
    name,
    title: nonEmpty(heading) ?? name,
    summary,
    contextTokens: estimateTokens(`${name}\n${summary}`),
    content: raw
  };
}

export function estimateTokens(text: string): number {
  const scalarCount = [...text].length;
  const wordCount = text.split(/[\s\p{Punctuation}]+/u).filter(Boolean).length;
  const cjkCount = [...text].filter((char) => {
    const code = char.codePointAt(0) ?? 0;
    return code >= 0x4e00 && code <= 0x9fff;
  }).length;

  const latinEstimate = Math.max(wordCount, Math.floor(scalarCount / 5));
  return Math.max(1, latinEstimate + Math.floor(cjkCount / 2));
}

function extractFrontmatter(raw: string): Record<string, string> {
  const lines = raw.split(/\r?\n/);
  if (lines[0]?.trim() !== "---") return {};

  const result: Record<string, string> = {};
  for (let index = 1; index < lines.length; index += 1) {
    const line = lines[index] ?? "";
    if (line.trim() === "---") break;
    const match = /^([A-Za-z0-9_-]+):\s*(.*)$/.exec(line);
    if (!match) continue;
    const key = match[1]!.trim();
    let value = match[2]!.trim();
    if (value === ">" || value === "|") {
      const blockLines: string[] = [];
      for (index += 1; index < lines.length; index += 1) {
        const blockLine = lines[index] ?? "";
        if (blockLine.trim() === "---") {
          index -= 1;
          break;
        }
        if (/^[A-Za-z0-9_-]+:\s*/.test(blockLine)) {
          index -= 1;
          break;
        }
        blockLines.push(blockLine.replace(/^\s{2,}/, "").trimEnd());
      }
      value = value === ">" ? blockLines.map((item) => item.trim()).filter(Boolean).join(" ") : blockLines.join("\n").trim();
    }
    if (value.startsWith('"') && value.endsWith('"') && value.length >= 2) {
      value = value.slice(1, -1);
    }
    result[key] = value;
  }
  return result;
}

function firstMeaningfulParagraph(raw: string): string | null {
  for (const paragraph of raw.split(/\n\s*\n/)) {
    const trimmed = paragraph.trim();
    if (!trimmed) continue;
    if (trimmed.startsWith("---")) continue;
    if (trimmed.startsWith("#")) continue;
    if (trimmed.length <= 16) continue;
    return trimmed;
  }
  return null;
}

function nonEmpty(value: string | null | undefined): string | null {
  const trimmed = value?.trim();
  return trimmed ? trimmed : null;
}
